### PR TITLE
Implement flip and flipWith

### PR DESCRIPTION
**Pull request by Executioner1939**
#41 

@Avasil I think my examples will need some looking at, I really couldn't think of anything better at the time, perhaps we can come up with some better ones?

### CODE REVIEW

File: core/shared/src/main/scala/monix/bio/BIO.scala [hunk 1]
```diff
@@ -1788,6 +1788,64 @@ sealed abstract class BIO[+E, +A] extends Serializable {
...
+    *
+    *   f1.flipWith(_.flatMap(sendHttpResponse(_))) // Respond to client as soon as possible.
+    *     .flatMap(logToStdout(_))    // Perform actions after response has been sent.
+    *     .flatMap(logErrorToFile(_))
+    *
+    * }}}
+    *
+    * @param f function to apply to the flipped error value.
+    */
+  final def flipWith[E1, A1](f: BIO[A, E] => BIO[E1, A1]): BIO[E1, A1] = {
```
Comment (Executioner1939) [inline core/shared/src/main/scala/monix/bio/BIO.scala:1845]: @Avasil I see in ZIO they have a different signature:

`final def flipWith[E1, A1](f: BIO[A, E] => BIO[E1, A1]): BIO[A1, E1]`

Comment (Avasil) [reply core/shared/src/main/scala/monix/bio/BIO.scala:1845]: Oh, I think it should be:

```scala
final def flipWith[E1, A1](f: BIO[A, E] => BIO[A1, E1]): BIO[E1, A1]
```

Comment (Executioner1939) [reply core/shared/src/main/scala/monix/bio/BIO.scala:1845]: Awesome, will update this 👍 as well as add a better example and some more tests.

Comment (Executioner1939) [reply core/shared/src/main/scala/monix/bio/BIO.scala]: Hi :-). I have updated the Tests and Docs, please let me know if I can do anything else :-)

File: core/shared/src/main/scala/monix/bio/BIO.scala [hunk 2]
```diff
@@ -1788,6 +1788,70 @@ sealed abstract class BIO[+E, +A] extends Serializable {
...
+    *   } yield ()
+    *
+    * }}}
+    */
+  final def flip: BIO[A, E] = {
+    this.redeemWith(BIO.now, BIO.raiseError)
+  }
+
+  /***
+    * This function is a useful alternative to [[flip]] in that it applies a series
```
Comment (Avasil) [inline core/shared/src/main/scala/monix/bio/BIO.scala:1825]: I would say it's more of a helper which implements a common pattern with `flip` than alternative to it

File: core/shared/src/main/scala/monix/bio/BIO.scala [hunk 3]
```diff
@@ -1788,6 +1788,70 @@ sealed abstract class BIO[+E, +A] extends Serializable {
   final def failed: UIO[E] =
     FlatMap(this, BIO.Failed.asInstanceOf[StackFrame[E, A, UIO[E]]])
 
+  /**
+    * Creates a new BIO by swapping the error and value parameters.
+    *
+    * This allows you to work with the error in the right-biased context
+    * and allows for easier error handling and chaining of events that
```
Comment (Avasil) [inline core/shared/src/main/scala/monix/bio/BIO.scala:1795]: What do you mean by easier error handling?

File: core/shared/src/main/scala/monix/bio/BIO.scala (hunk 2 above)
Comment (Executioner1939) [reply core/shared/src/main/scala/monix/bio/BIO.scala:1825]: Thanks :-). Will change the wording :-)

File: core/shared/src/main/scala/monix/bio/BIO.scala (hunk 3 above)
Comment (Executioner1939) [reply core/shared/src/main/scala/monix/bio/BIO.scala:1795]: I need to reword this, what I am trying to say is it's easier to work with the error type in a right biased context using map, flatMap.

**Comment by Executioner1939** [comment]
> Thanks, I really appreciate scaladocs!

My pleasure, really tried to think of a nice example. Was quite tough to be honest 🙈 🙈

File: core/shared/src/test/scala/monix/bio/TaskFlipSuite.scala [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+ * limitations under the License.
+ */
+
+package monix.bio
+
+import monix.execution.exceptions.DummyException
+
+import scala.util.{Failure, Success}
+
+object TaskFlipSuite extends BaseTestSuite {
```
[the remark this answers is not in the document]
Comment (Executioner1939) [reply core/shared/src/test/scala/monix/bio/TaskFlipSuite.scala:24]: I will implement these :-)

**Comment by Executioner1939** [comment]
@Avasil my pleasure, I was also struggling to come up with some use cases. If you think of something really awesome I would love to see :-).

**Pull request merged**